File: problems/746-Min-Cost-Climbing-Stairs/solution.go
func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func minCostClimbingStairs(cost []int) int {
	dp := make([]int, 2)

	for i := 2; i < len(cost); i++ {
		next := min(dp[0]+cost[i-2], dp[1]+cost[i-1])
		dp[0] = dp[1]
		dp[1] = next
	}

	return min(dp[0]+cost[len(cost)-2], dp[1]+cost[len(cost)-1])
}

File: problems/746-Min-Cost-Climbing-Stairs/solution.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int[] dp = new int[cost.length];
        dp[dp.length-1] = cost[cost.length-1];
        dp[dp.length-2] = cost[cost.length-2];

        for (int i = dp.length-3; i >= 0; i--) {
            dp[i] = cost[i] + Math.min(dp[i+1], dp[i+2]);
        }

        return Math.min(dp[0], dp[1]);
    }
}

